File: RenetikObjc/Library/Classes/Core/Categories/UIKit/UITextView+CSExtension.m
#import "UITextView+CSExtension.h"
#import "CSLang.h"
#import "NSString+CSExtension.h"
#import "NSObject+CSExtension.h"
#import "UIView+CSExtension.h"
#import "UIView+CSDimension.h"

@implementation UITextView (CSExtension)

- (instancetype)setHTMLFromString:(NSString *)string {
    string = [string add:stringf(
            @"<style>body{font-family: '%@'; font-size:%fpx;}</style>", self.font.fontName, self.font.pointSize)];
    self.attributedText = [NSAttributedString.alloc initWithData:[string dataUsingEncoding:NSUnicodeStringEncoding] options:@{NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType, NSCharacterEncodingDocumentAttribute: @(NSUTF8StringEncoding)}
                                              documentAttributes:nil error:nil];
    return self;
}

- (instancetype)htmlByString:(NSString *)string {
    [self setHTMLFromString:string];
    return self;
}

+ (void)hideTextInsets:(NSArray<UITextView *> *)textViews {
    for (UITextView *view in textViews) view.textContainerInset = UIEdgeInsetsZero;
}

- (instancetype)asLabel {
    self.textContainerInset = UIEdgeInsetsZero;
    self.contentInset = UIEdgeInsetsZero;
    self.editable = false;
    self.scrollEnabled = false;
    self.backgroundColor = UIColor.clearColor;
    if (@available(iOS 11, *))
        self.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    self.textContainer.lineFragmentPadding = 0;
    self.layoutManager.usesFontLeading = false;
    return self;
}

- (instancetype)scrollToCursorPosition {
    doLater(0.1, ^{
        [self scrollRectToVisible:[self caretRectForPosition:self.selectedTextRange.start] animated:YES];
    });
    return self;
}

- (instancetype)dataDetector:(UIDataDetectorTypes)types {
    self.dataDetectorTypes = types;
    if (types != UIDataDetectorTypeNone) self.selectable = true;
    return self;
}

- (instancetype)text:(NSString *)text {
    self.text = text;
    return self;
}

- (instancetype)alignText:(enum NSTextAlignment)alignment {
    self.textAlignment = alignment;
    return self;
}

- (instancetype)font:(UIFont *)font {
    self.font = font;
    return self;
}

- (instancetype)textColor:(UIColor *)textColor {
    self.textColor = textColor;
    return self;
}

- (instancetype)setFontSize:(CGFloat)size {
    self.font = [self.font fontWithSize:size];
    return self;
}

- (instancetype)fontSize:(CGFloat)size {
    self.font = [self.font fontWithSize:size];
    return self;
}

- (void)setFontStyle:(UIFontTextStyle)style {
    self.font = [UIFont preferredFontForTextStyle:style];
}

- (instancetype)fontStyle:(UIFontTextStyle)style {
    self.fontStyle = style;
    return self;
}

- (instancetype)heightToLines:(NSInteger)numberOfLines {
    let currentWidth = self.width;
    let currentText = self.text;
    self.text = @"line";
    for (int i = 0; i < numberOfLines - 1; i++) self.text = [self.text add:@"\n line"];
    [self sizeToFit];
    self.text = currentText;
    self.width = currentWidth;
    return self;
}

- (instancetype)onClick:(void (^)(UIView *))block {
    self.editable = false;
    self.selectable = false;
    return [super onClick:block];
}

@end

File: Renetik/Renetik/Classes/Core/Extensions/UIKit/UITextView+CSExtension.swift
//
// Created by Rene Dohan on 1/25/20.
//

import UIKit
import RenetikObjc
import BlocksKit

public struct CSTextViewClearButtonAppearance {
    fileprivate let titleColor: UIColor?

    public init(titleColor: UIColor) {
        self.titleColor = titleColor
    }
}

public extension UITextView {

    class func construct(_ text: String) -> Self { construct().text(text) }

    func delegate(_ delegate: UITextViewDelegate) -> Self { self.delegate = delegate; return self }

    func scrollToCursorPosition() -> Self {
        later(seconds: 0.1) {
            if let start = self.selectedTextRange?.start {
                self.scrollRectToVisible(self.caretRect(for: start), animated: true)
            }
        }
        return self
    }

    @discardableResult
    func detectData(_ types: UIDataDetectorTypes = [.all]) -> Self {
        dataDetectorTypes = types
        if types.isEmpty { isSelectable = true }
        return self
    }

    @discardableResult
    func detect(data: UIDataDetectorTypes) -> Self { detectData(data) }

    @discardableResult
    func detect(_ data: UIDataDetectorTypes) -> Self { detectData(data) }

    @discardableResult
    override public func onTap(_ block: @escaping Func) -> Self {
        self.isEditable = false
        self.isSelectable = false
        super.onTap(block)
        return self
    }

//    TODO: This is wrong and TextView should be wrapped to view with clear button instead
//    @discardableResult
//    func withClear(_ parent: CSViewController, _ appearance: CSTextViewClearButtonAppearance? = nil) -> Self {
//        let button = UIButton(type: .system).construct().text("X").fontStyle(.body)
//                .visible(if: self.text.isSet).onClick { self.text = "" }
//        add(button).from(left: 5).resizeToFit().centeredHorizontal()
//
//        onTextChange(in: parent) { _ in button.visible(if: self.text.isSet) }
//        appearance?.titleColor?.then { color in button.textColor(color) }
//        return self
//    }

    @discardableResult
    public func text(_ value: String?) -> Self { invoke { self.text = value } }

    @discardableResult
    func onTextChange(in parent: CSViewController, _ function: @escaping (UITextView) -> Void) -> Self {
        parent.observe(notification: UITextView.textDidChangeNotification) { notification in
            if (notification.object as! NSObject) === self { function(self) }
        }
        return self
    }

    @discardableResult
    func heightToFit(lines numberOfLines: Int) -> Self {
        let currentWidth = width; let currentText = text; var linesText = "line"
        for i in 0..<numberOfLines - 1 {
            linesText += "\n line"
        }
        text(linesText).resizeToFit().text(currentText).width(currentWidth)
        return self
    }

    @discardableResult
    func asLabel() -> Self {
        textContainerInset = .zero
        contentInset = .zero
        isEditable = false
        isScrollEnabled = false
        backgroundColor = .clear
        contentInsetAdjustmentBehavior = .never
        textContainer.lineFragmentPadding = 0
        layoutManager.usesFontLeading = false
        return self
    }

    @discardableResult
    func text(align: NSTextAlignment) -> Self { invoke { self.textAlignment = align } }

    @discardableResult
    func font(_ font: UIFont) -> Self { invoke { self.font = font } }

    @discardableResult
    func font(size: CGFloat) -> Self { invoke { self.fontSize = size } }

    var fontSize: CGFloat {
        get { font!.fontDescriptor.pointSize }
        set { font = font!.withSize(newValue) }
    }

    @discardableResult
    func font(style: UIFont.TextStyle) -> Self { invoke { self.fontStyle = style } }

    var fontStyle: UIFont.TextStyle {
        get { font!.fontDescriptor.object(forKey: .textStyle) as! UIFont.TextStyle }
        set { font = UIFont.preferredFont(forTextStyle: newValue) }
    }

    @discardableResult
    public func text(prepend: String) -> Self { self.text("\(prepend)\(text ?? "")") }

    @discardableResult
    public func text(replace: String, with: String) -> Self { self.text(text.replace(all: replace, with: with)) }
}